client.go

```diff
@@ -66,85 +66,130 @@ func (c *Client) Close() error {
+// retrieving the responses afterwards.
+func (c *Client) Request(ip net.IP) error {
 	// Create ARP packet for broadcast address to attempt to find the
 	// hardware address of the input IP address
 	arp, err := NewPacket(OperationRequest, c.ifi.HardwareAddr, c.ip, ethernet.Broadcast, ip)
 	if err != nil {
-		return nil, err
-	}
-	arpb, err := arp.MarshalBinary()
-	if err != nil {
-		return nil, err
-	}
-
-	// Create ethernet frame addressed to broadcast address to encapsulate the
-	// ARP packet
-	eth := &ethernet.Frame{
-		Destination: ethernet.Broadcast,
-		Source:      c.ifi.HardwareAddr,
-		EtherType:   ethernet.EtherTypeARP,
-		Payload:     arpb,
-	}
-	ethb, err := eth.MarshalBinary()
-	if err != nil {
-		return nil, err
+		return err
 	}
+	return c.WriteTo(arp, ethernet.Broadcast)
+}
 
-	// Write frame to ethernet broadcast address
-	_, err = c.p.WriteTo(ethb, &raw.Addr{
-		HardwareAddr: ethernet.Broadcast,
-	})
+// Resolve performs an ARP request, attempting to retrieve the
+// hardware address of a machine using its IPv4 address. Resolve must not
+// be used concurrently with Read. If you're using Read (usually in a
+// loop), you need to use Request instead. Resolve may read more than
+// one message if it receives messages unrelated to the request.
+func (c *Client) Resolve(ip net.IP) (net.HardwareAddr, error) {
+	err := c.Request(ip)
 	if err != nil {
 		return nil, err
 	}
 
 	// Loop and wait for replies
-	buf := make([]byte, 128)
 	for {
-		n, _, err := c.p.ReadFrom(buf)
+		arp, eth, err := c.Read()
 		if err != nil {
 			return nil, err
 		}
 
-		// Unmarshal ethernet frame and check:
-		//   - Frame is for our hardware address
-		//   - Frame has ARP EtherType
-		if err := eth.UnmarshalBinary(buf[:n]); err != nil {
-			return nil, err
-		}
 		if !bytes.Equal(eth.Destination, c.ifi.HardwareAddr) {
 			continue
 		}
-		if eth.EtherType != ethernet.EtherTypeARP {
+		if !c.IsReplyFrom(ip, arp) {
 			continue
 		}
 
-		// Unmarshal ARP packet and check:
-		//   - Packet is a reply, not a request
-		//   - Packet is for our IP address
-		//   - Packet is for our hardware address
-		//   - Packet is a reply to our query, not another query
-		if err := arp.UnmarshalBinary(eth.Payload); err != nil {
-			return nil, err
-		}
-		if arp.Operation != OperationReply {
-			continue
-		}
-		if !arp.TargetIP.Equal(c.ip) {
-			continue
+		return arp.SenderHardwareAddr, nil
+	}
+}
+
+// IsReply reports whether the packet p is a response to one of our
+// queries. A packet is a response if it is addressed to our IP- and
+// hardware address.
+func (c *Client) IsReply(p *Packet) bool {
+	return p.Operation == OperationReply &&
+		p.TargetIP.Equal(c.ip) &&
+		bytes.Equal(p.TargetHardwareAddr, c.ifi.HardwareAddr)
+}
+
+// IsReplyFrom reports whether the packet p is a response to our query
+// for ip. It works like IsReply, but additionally checks if the
+// response is for a specific IPv4 address ip.
```

Comment:
Should "for" be "from" here?  The code seems to indicate it's looking at `p.SenderIP`.